chia/plot_sync/receiver.py

```diff
@@ -0,0 +1,313 @@
+from chia.plot_sync.exceptions import (
+    InvalidIdentifierError,
+    InvalidLastSyncIdError,
+    PlotAlreadyAvailableError,
+    PlotNotAvailableError,
+    PlotSyncException,
+    SyncIdsMatchError,
+)
+from chia.plot_sync.util import ErrorCodes, State
+from chia.protocols.harvester_protocol import (
+    Plot,
+    PlotSyncDone,
+    PlotSyncError,
+    PlotSyncIdentifier,
+    PlotSyncPathList,
+    PlotSyncPlotList,
+    PlotSyncResponse,
+    PlotSyncStart,
+)
+from chia.server.ws_connection import ProtocolMessageTypes, WSChiaConnection, make_msg
+from chia.util.ints import int16, uint64
+
+log = logging.getLogger(__name__)
+
+
+class Receiver:
+    _connection: WSChiaConnection
+    _sync_state: State
+    _delta: Delta
+    _expected_sync_id: uint64
+    _expected_message_id: uint64
+    _last_sync_id: uint64
+    _last_sync_time: float
+    _plots: Dict[str, Plot]
+    _invalid: List[str]
+    _keys_missing: List[str]
+    _duplicates: List[str]
+    _update_callback: Callable
+
+    def __init__(self, connection: WSChiaConnection, update_callback: Callable) -> None:
+        self._connection = connection
+        self._sync_state = State.idle
+        self._delta = Delta()
+        self._expected_sync_id = uint64(0)
+        self._expected_message_id = uint64(0)
+        self._last_sync_id = uint64(0)
+        self._last_sync_time = 0
+        self._plots = {}
+        self._invalid = []
+        self._keys_missing = []
+        self._duplicates = []
+        self._update_callback = update_callback  # type: ignore[assignment]
+
+    def reset(self) -> None:
+        self._sync_state = State.idle
+        self._expected_sync_id = uint64(0)
+        self._expected_message_id = uint64(0)
+        self._last_sync_id = uint64(0)
+        self._last_sync_time = 0
+        self._plots.clear()
+        self._invalid.clear()
+        self._keys_missing.clear()
+        self._duplicates.clear()
+        self._delta.clear()
+
+    def bump_expected_message_id(self) -> None:
+        self._expected_message_id = uint64(self._expected_message_id + 1)
+
+    def connection(self) -> WSChiaConnection:
+        return self._connection
+
+    def state(self) -> State:
+        return self._sync_state
+
+    def expected_sync_id(self) -> uint64:
+        return self._expected_sync_id
+
+    def expected_message_id(self) -> uint64:
+        return self._expected_message_id
+
+    def last_sync_id(self) -> uint64:
+        return self._last_sync_id
+
+    def last_sync_time(self) -> float:
+        return self._last_sync_time
+
+    def plots(self) -> Dict[str, Plot]:
+        return self._plots
+
+    def invalid(self) -> List[str]:
+        return self._invalid
+
+    def keys_missing(self) -> List[str]:
+        return self._keys_missing
+
+    def duplicates(self) -> List[str]:
+        return self._duplicates
+
+    async def _process(self, method: Callable, message_type: ProtocolMessageTypes, message: Any) -> None:
```

Comment:
Hmm not too bad, i guess the benefit you see is to avoid the `_methods` ? But imo it also adds a bit more obscurity to the thing (i mean, the callback is kind of straight forward and the context manager needs to be understood more)? Also this conflicts with https://github.com/Chia-Network/chia-blockchain/blob/pr-protocol-plot-sync/tests/plot_sync/test_sync_simulated.py#L229 then it seems?